File: backend/app/api/routes/health.py
"""Health check endpoints — liveness and readiness."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from starlette import status

from backend.app.dependencies import get_readiness_service, get_settings
from backend.app.infrastructure.config import Settings
from backend.app.infrastructure.health.models import HealthResponse, ReadinessResponse
from backend.app.infrastructure.health.service import ReadinessService

router: APIRouter = APIRouter(tags=["health"])


@router.get("/health", response_model=HealthResponse)
async def health(settings: Settings = Depends(get_settings)) -> HealthResponse:
    """Liveness check.

    Returns 200 if the application process is alive.
    Does NOT check external dependencies.
    """
    return HealthResponse(
        status="ok",
        service=settings.app_name,
        version=settings.app_version,
    )


@router.get(
    "/ready",
    response_model=ReadinessResponse,
    responses={503: {"model": ReadinessResponse}},
)
async def readiness(
    readiness: ReadinessService = Depends(get_readiness_service),
) -> ReadinessResponse:
    """Readiness check.

    Returns 200 with status=ready when all mandatory dependencies are healthy.
    Returns 503 with status=not_ready when any dependency is unavailable.
    The HTTP status code mirrors the overall readiness state.
    """
    result = await readiness.check_readiness()

    if result.status.value == "not_ready":
        return JSONResponse(  # type: ignore[return-value]
            content=result.model_dump(),
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
        )

    return result

File: desktop/src/api/services/health.ts
/**
 * Health Service - API service for health and readiness endpoints.
 *
 * Provides typed access to the FastAPI health check endpoints.
 */

import { httpClient } from "../client";
import type { HealthResponse, ReadinessResponse } from "../types";

/**
 * Health check - liveness probe.
 * Returns 200 if the application process is alive.
 * Does NOT check external dependencies.
 */
export async function getHealth(): Promise<HealthResponse> {
  const response = await httpClient.get<HealthResponse>("/health");
  return response.data;
}

/**
 * Readiness check - checks all mandatory dependencies.
 * Returns 200 with status=ready when all dependencies are healthy.
 * Returns 503 with status=not_ready when any dependency is unavailable.
 */
export async function getReadiness(): Promise<ReadinessResponse> {
  const response = await httpClient.get<ReadinessResponse>("/ready");
  return response.data;
}

/**
 * Check if the API is healthy and ready.
 * Returns true if both health and readiness are ok.
 */
export async function isApiReady(): Promise<boolean> {
  try {
    const [health, readiness] = await Promise.all([getHealth(), getReadiness()]);
    return health.status === "ok" && readiness.status === "ready";
  } catch {
    return false;
  }
}

/**
 * Get API version info from health endpoint.
 */
export async function getApiVersion(): Promise<{ service: string; version: string } | null> {
  try {
    const response = await httpClient.get<HealthResponse>("/health");
    return {
      service: response.data.service,
      version: response.data.version,
    };
  } catch {
    return null;
  }
}

export { getHealth as healthCheck, getReadiness as readinessCheck };
export type { HealthResponse, ReadinessResponse };
